core/unix/os.c

```diff
@@ -6988,12 +6988,21 @@ pre_system_call(dcontext_t *dcontext)
                  * thread in the post-syscall handling of clone3; as it is used
                  * only by the parent thread, there is no use-after-free danger here.
                  */
-                clone3_syscall_args_t *dr_clone_args = HEAP_TYPE_ALLOC(
-                    dcontext, clone3_syscall_args_t, ACCT_OTHER, PROTECTED);
+                uint app_clone_args_size =
+                    (uint)sys_param(dcontext, SYSCALL_PARAM_CLONE3_CLONE_ARGS_SIZE);
+                /* Note that the struct clone_args being used by the app may have
+                 * more fields than DR's internal struct (clone3_syscall_args_t).
+                 * For creating DR's copy of the app's clone_args object, we need to
+                 * allocate as much space as specified by the app in the clone3
+                 * syscall's args.
+                 */
+                clone3_syscall_args_t *dr_clone_args =
+                    (clone3_syscall_args_t *)heap_alloc(
+                        dcontext, app_clone_args_size HEAPACCT(ACCT_OTHER));
                 clone3_syscall_args_t *app_clone_args =
                     (clone3_syscall_args_t *)sys_param(dcontext,
                                                        SYSCALL_PARAM_CLONE3_CLONE_ARGS);
-                memcpy(dr_clone_args, app_clone_args, sizeof(clone3_syscall_args_t));
+                memcpy(dr_clone_args, app_clone_args, app_clone_args_size);
```

Comment:
As discussed offline, let's make this a d_r_safe_read() to avoid DR crashing if an app passes bogus memory to the kernel.

Reply:
Done.